### DISCUSSION TITLE
Mapping one source to several targets, problems Feedback and Relative

### DISCUSSION THREAD

**Discussion by vonglan** [helgoboss/helgobox#293] [General]
In the Analog Synth topic, it is sometimes useful to be able to map one source to multiple targets (depending on the VST, so this is in the Main Mapping):
A) Continuous Waveform to Sawtooth On/Off, Triangle On/Off, Pulse On/Off
B) Bipolar Modulation amount to Unipolar Modulation Amount+Inversion Switch 

I just realized, that this is problematic when you also want to use Feedback or Relative sources.

Feedback:
In theory this would be possible (if ReaLearn offered access to multiple FX target parameters in the Feedback Transformation EEL formula), but I think it is too much implementation effort.

Relative:
The Control formulas for A and B would not work on relativ values. So that is what the „Make absolute“ setting is for!
Would it be possible to use the „internal absolute value“ described in the User’s Guide for the Feedback?

Thinking about decoupling: The Controller Mapping should not have to guess whether the Main Mapping uses an EEL formula for this source or not. Instead, it would be good if "make absolute" was automatically performed in the Controller layer, in the case
- Source element in the Controller Mapping is relative
- In the Main Mapping, an EEL formula for Control transformation is used for the related virtual source
(although maybe people also want to use an EEL formula on relative values in some cases?)

**Comment by helgoboss** [comment]
> In the Analog Synth topic, it is sometimes useful to be able to map one source to multiple targets (depending on the VST, so this is in the Main Mapping):
> A) Continuous Waveform to Sawtooth On/Off, Triangle On/Off, Pulse On/Off
> B) Bipolar Modulation amount to Unipolar Modulation Amount+Inversion Switch
> 
> I just realized, that this is problematic when you also want to use Feedback or Relative sources.
> 
> Feedback:
> In theory this would be possible (if ReaLearn offered access to multiple FX target parameters in the Feedback Transformation EEL formula), but I think it is too much implementation effort.

If you have multiple mappings mapped to one control element, you need to disable feedback for all mappings except one. #4 will make it possible to have multiple targets (but only based on wildcards and `<Selected>`) in one mapping and maybe it will calculate some kind of average feedback value.

> 
> Relative:
> The Control formulas for A and B would not work on relativ values. So that is what the „Make absolute“ setting is for!
> Would it be possible to use the „internal absolute value“ described in the User’s Guide for the Feedback?
> 
> Thinking about decoupling: The Controller Mapping should not have to guess whether the Main Mapping uses an EEL formula for this source or not. Instead, it would be good if "make absolute" was automatically performed in the Controller layer, in the case
> * Source element in the Controller Mapping is relative
> * In the Main Mapping, an EEL formula for Control transformation is used for the related virtual source
>   (although maybe people also want to use an EEL formula on relative values in some cases?)

Exactly, the controller mapping should not have to guess, it should not decide about it. If you have an encoder or some relative control element, you should really leave it at that in the controller mappings. Don't make it absolute already! The main mapping should use "Make absolute" if it insists on applying an EEL formula even for relative values, not the controller mapping.

**Reply by vonglan** [reply]
> maybe it will calculate some kind of average feedback value

No good for my use case.

> Exactly, the controller mapping should not have to guess, it should not decide about it. If you have an encoder or some relative control element, you should really leave it at that in the controller mappings. Don't make it absolute already! The main mapping should use "Make absolute" if it insists on applying an EEL formula even for relative values, not the controller mapping.

Ah, so this is possible already.

What about this question
> Would it be possible to use the „internal absolute value“ described in the User’s Guide for the Feedback?

**Reply by helgoboss** [reply]
> > maybe it will calculate some kind of average feedback value
> 
> No good for my use case.

How exactly do you think the feedback must work in your case? If you have one LED, how is it even possible to encode a little bit of Sawtooth On/Off, a little bit of Triangle On/Off and a little bit of Pulse On/Off into it? Through different colors? This must be cleared up at first.

> 
> > Exactly, the controller mapping should not have to guess, it should not decide about it. If you have an encoder or some relative control element, you should really leave it at that in the controller mappings. Don't make it absolute already! The main mapping should use "Make absolute" if it insists on applying an EEL formula even for relative values, not the controller mapping.
> 
> Ah, so this is possible already.
> 
> What about this question
> 
> > Would it be possible to use the „internal absolute value“ described in the User’s Guide for the Feedback?

You can access it in the control transformation formula but not the feedback transformation formula. It wouldn't make sense because feedback flows from targets to sources, not from sources to targets. Also note that feedback is *always* absolute, even with relative control elements.

**Reply by vonglan** [reply]
> How exactly do you think the feedback must work in your case? If you have one LED, how is it even possible to encode a little bit of Sawtooth On/Off, a little bit of Triangle On/Off and a little bit of Pulse On/Off into it? Through different colors? This must be cleared up at first.

I am sure it is too much effort and not worth it for this special case, but if you are interested:
I convert the waveshape range 0-127 to a series of 000,001,010,011, etc. for the three on/off targets (like binary counting). So for the Feedback, you could write a formula: waveshape feedback = sawtooth/2 + triangle/4 + pulse/8.

> > > Would it be possible to use the „internal absolute value“ described in the User’s Guide for the Feedback?
> 
> You can access it in the control transformation formula but not the feedback transformation formula. It wouldn't make sense because feedback flows from targets to sources, not from sources to targets. Also note that feedback is _always_ absolute, even with relative control elements.

Even if it does not generally make sense, I think it would be a perfect and elegant solution for feedback in case A. 
As I have currently implemented it (feedback switched off), the X-Touch will just „guess“ the absolute value. Which works relatively (;-)) good.

For case B, I just leave feedback on for the target „Unipolar modulation amount“. That already works nicely. With the X-Touch (relative mode), I can only access (control) the currently selected (positive or negative) half of the range, but that is completely acceptable.
(Because this is conceived as a standard set of virtual source/targets, it will always be lacking in a few special areas, where a specialized mapping e.g. X-Touch to Repro-5 could be more complete or a little bit better.)

**Comment by vonglan** [comment]
Another problem appeared (and is solved - see last two sentences): for mapping a bipolar modulation amount to unipolar plus an "inverse" switch, I use the following EEL formulas:

VST Parameter | EEL Control Transformation | Comment
-- | -- | --
Unipolar Modulation Amount | y=2*abs(x-0.5) | Leave Feedback on
Invert Switch  | y = (x == 0.5) ? 0 : 1-x | in neutral position, invert should be off

I do not use "make absolute".

This works fine with absolute (OB6) or relative (X-Touch) Controller mappings, for u-he Repro.
With the absolute controller, I can control both the negative and the positive half of the Modulation range.
With the relative controller, I stay in whatever half is currently selected. The "Invert switch" is not changed (which is what I want).

However, the same mappings (duplicated, then only targets changed) do not work for TAL J-8: In this case, when I use the relative controller, the "Invert switch" is switched on and off, when the (absolute) unipolar modulation amount exceeds certain values (depending on history).
When I display the FX parameters in the Track Control, I see that u-he displays the Polarity as "Positive" or "Negative". Whereas TAL uses values of 0 to 1. So if I use the OB6 first, and turn the bipolar amount control to the neutral position, TAL shows "0.5" as parameter value. And when I use the X-Touch encoder after that, it starts from those 0.5 (and quickly switches on "invert"). 
Is there a way to tell ReaLearn in the (J-8 Main) mapping: "ignore input if the source is relative"? Ah - I can set Step size min=max=0. That seems to solve it!
